codegen/builtin_fs.go

```diff
@@ -0,0 +1,499 @@
+
+	localOpts = append(localOpts, llb.SharedKeyHint(localPath), llb.WithDescription(map[string]string{
+		solver.LocalPathDescriptionKey: fmt.Sprintf("local://%s", localPath),
+	}))
+
+	sessionID := SessionID(ctx)
+	if sessionID != "" {
+		localOpts = append(localOpts, llb.SessionID(sessionID))
+	}
+
+	fs := Filesystem{State: llb.Local(id, localOpts...)}
+	fs.SessionOpts = append(fs.SessionOpts, WithSyncedDir(id, filesync.SyncedDir{
+		Name: id,
+		Dir:  localPath,
+		Map: func(_ string, st *fstypes.Stat) bool {
+			st.Uid = 0
+			st.Gid = 0
+			return true
+		},
+	})
+
+	return ret.Set(fs)
+}
+
+type Frontend struct{}
+
+func (f Frontend) Call(ctx context.Context, ret Value, opts Options, source, ref string) error {
+	var gatewayOpts []llbutil.GatewayOption
+	for _, opt := range opts {
+		gatewayOpts = append(gatewayOpts, opt.(llbutil.GatewayOption))
+	}
+
+	st = llb.Scratch().Async(func(ctx context.Context, ret Value, opts Options) (llb.State, error) {
+		// FIXME: solver Build
+	})
+
+	return ret.Set(Filesystem{
+		State: st,
+		Image: nil, // FIXME: proxy image
+	})
+}
+
+type Env struct{}
+
+func (e Env) Call(ctx context.Context, ret Value, opts Options, key, value string) error {
+	fs, err := ret.AsFilesystem()
+	if err != nil {
+		return err
+	}
+
+	fs.State = fs.State.AddEnv(key, value)
+	return ret.Set(fs)
+}
+
+type Dir struct{}
+
+func (d Dir) Call(ctx context.Context, ret Value, opts Options, path string) error {
+	fs, err := ret.AsFilesystem()
+	if err != nil {
+		return err
+	}
+
+	fs.State = fs.State.Dir(path)
+	return ret.Set(fs)
+}
+
+type User struct{}
+
+func (u User) Call(ctx context.Context, ret Value, opts Options, name string) error {
+	fs, err := ret.AsFilesystem()
+	if err != nil {
+		return err
+	}
+
+	fs.State = fs.State.User(name)
+	return ret.Set(fs)
+}
+
+type Run struct{}
+
+func (r Run) Call(ctx context.Context, ret Value, opts Options, args ...string) error {
+	var (
+		runOpts []llb.RunOption
+		solveOpts []solver.SolveOption
+		sessionOpts []SessionOption
+		shlex   = false
+	)
+	for _, opt := range opts {
+		switch o := opt.(type) {
+		case llb.RunOption:
+			runOpts = append(runOpts, o)
+		case solver.SolveCallback:
+			solveOpts = append(solveOpts, o)
+		case SessionOption:
+			sessionOpts = append(sessionOpts, o)
+		case *Shlex:
+			shlex = true
+		}
+	}
+	for _, opt := range SourceMap(ctx) {
+		runOpts = append(runOpts, opt)
+	}
+
+	runArgs, err := ShlexArgs(args, shlex)
+	if err != nil {
+		return err
+	}
+
+	customName := strings.ReplaceAll(shellquote.Join(runArgs...), "\n", "")
+	runOpts = append(runOpts, llb.Args(runArgs), llb.WithCustomName(customName))
+
+	mounts := make(map[string]*parser.CallStmt)
+	// FIXME: populate mounts
+
+	err = llbutil.ShimReadonlyMountpoints(opts)
+	if err != nil {
+		return err
+	}
+
+	fs, err := ret.AsFilesystem()
+	if err != nil {
+		return err
+	}
+
+	run := fs.State.Run(runOpts...)
+	for path, mount := range mounts {
+		mnt := run.GetMount(path)
+		// FIXME: set bind
```

Comment:
How can this set the binding value? Where are the function registers defined? Maybe we need to pass a `CallingContext` or whatever into `Call()` to capture the return value and bind registers?

Reply:
The latest commit brings binding values in. I wanted to bring it closer to the concept that bind clauses are just an alias for a function but changes the target for the return value. In the current iteration, binds is basically used as a switch case to decide what value to write into `ret Register`.

I think caching the values returned from bindings should be generalized to having a `InstructionCache` that caches exprs in general.